src/Workspaces/Remote/Core/Diagnostics/PerformanceTrackerService.cs

```diff
@@ -0,0 +1,445 @@
+                    var kDistance = kDistances[rowIndex];
+
+                    for (var colIndex = 0; colIndex < analyzerCount; colIndex++)
+                    {
+                        var value = rowDistances[colIndex];
+
+                        // get neighbors closer than k distance
+                        if (value > 0 && value <= kDistance)
+                        {
+                            rowKNeighborIndices.Add(colIndex);
+                        }
+                    }
+
+                    kNeighborIndices[rowIndex] = rowKNeighborIndices;
+                }
+
+                return kNeighborIndices;
+            }
+
+            private List<double> GetKDistances(List<List<double>> allDistances, int k_value)
+            {
+                var analyzerCount = allDistances.Count;
+                var kDistances = GetPooledListAndSetCapacity<double>(analyzerCount);
+                var sortedRowDistance = GetPooledList<double>();
+
+                for (var index = 0; index < analyzerCount; index++)
+                {
+                    sortedRowDistance.Clear();
+                    sortedRowDistance.AddRange(allDistances[index]);
+
+                    sortedRowDistance.Sort();
+
+                    kDistances[index] = sortedRowDistance[k_value];
+                }
+
+                return kDistances;
+            }
+
+            private List<List<double>> GetAllDistances(List<(double normaliedAverage, double normalizedStddev)> normalizedMap)
+            {
+                var analyzerCount = normalizedMap.Count;
+                var allDistances = GetPooledListAndSetCapacity<List<double>>(analyzerCount);
+
+                for (var rowIndex = 0; rowIndex < analyzerCount; rowIndex++)
+                {
+                    var rowDistances = GetPooledListAndSetCapacity<double>(analyzerCount);
+                    var rowAnalyzer = normalizedMap[rowIndex];
+
+                    for (var colIndex = 0; colIndex < analyzerCount; colIndex++)
+                    {
+                        var colAnalyzer = normalizedMap[colIndex];
+                        var distance = Math.Sqrt(Math.Pow(colAnalyzer.normaliedAverage - rowAnalyzer.normaliedAverage, 2) +
+                                                 Math.Pow(colAnalyzer.normalizedStddev - rowAnalyzer.normalizedStddev, 2));
+
+                        rowDistances[colIndex] = distance;
+                    }
+
+                    allDistances[rowIndex] = rowDistances;
+                }
+
+                return allDistances;
+            }
+
+            private List<(double normaliedAverage, double normalizedStddev)> GetNormalizedPerformanceMap(
+                List<string> analyzerIdIndex, Dictionary<string, (double average, double stddev)> rawPerformanceData)
+            {
+                var (averageMin, averageMax) = (rawPerformanceData.Values.Select(kv => kv.average).Min(), rawPerformanceData.Values.Select(kv => kv.average).Max());
```

Comment:
❕ One declared variable per line (unless modified to track the min and max with a single iteration)